File: ctci/p131.py
def fib1(n):
    """Book solution from page 131."""
    if n == 0:
        return 0
    if n == 1:
        return 1
    return fib1(n - 1) + fib1(n - 2)


def fib2(n):
    """Own solution using memoization."""
    cache = dict()
    cache[0] = 0
    cache[1] = 1
    def _fib(n):
        if n in cache:
            return cache[n]
        cache[n] = _fib(n - 1) + _fib(n - 2)
        return cache[n]
    return _fib(n)


def fib3(n):
    """Own solution using memoization, slighly modified to match book."""
    cache = dict()
    cache[0] = 0
    cache[1] = 1
    def _fib(n):
        if n not in cache:
            cache[n] = _fib(n - 1) + _fib(n - 2)
        return cache[n]
    return _fib(n)


def fib4(n):
    """Book solution from page 133."""
    return _fib4(n, dict())


def _fib4(n, memo):
    if n == 0 or n == 1:
        return n
    if n not in memo:
        memo[n] = _fib4(n - 1, memo) + _fib4(n - 2, memo)
    return memo[n]


def fib5(n):
    """Bottom-up approach. Book solution from page 134."""
    memo = dict()
    memo[0] = 0
    memo[1] = 1
    for i in range(2, n + 1):
        memo[i] = memo[i - 1] + memo[i - 2]
    return memo[n]


def fib6(n):
    """Bottom-up approach that avoids caching entirely."""
    if n < 2:
        return n
    a, b = 0, 1
    for i in range(2, n + 1):
        c = a + b
        a, b = b, c
    return c


def fib7(n):
    """Bottom-up approach that avoids caching entirely. Book solution from
    page 134.
    """
    if n == 0:
        return n
    a, b = 0, 1
    for i in range(2, n):
        c = a + b
        a, b = b, c
    return a + b


impls = (fib1, fib2, fib3, fib4, fib5, fib6, fib7)

File: ctci/p131.cpp
#include <functional>

#include <gtest/gtest.h>

/**
 * Book solution from page 131.
 */
unsigned fib1(unsigned n) {
    if (n == 0)
        return 0;
    if (n == 1)
        return 1;
    return fib1(n - 1) + fib1(n - 2);
}

typedef std::map<unsigned, unsigned> cache_type;

unsigned fib2_helper(unsigned n, cache_type &cache) {
    if (cache.count(n) > 0)
        return cache.at(n);
    cache.insert({n, fib2_helper(n - 1, cache) + fib2_helper(n - 2, cache)});
    return cache.at(n);
}

/**
 * Own solution using memoization.
 */
unsigned fib2_nolambda(unsigned n) {
    cache_type cache;
    cache.insert({0, 0});
    cache.insert({1, 1});

    return fib2_helper(n, cache);
}

unsigned fib2(unsigned n) {
    cache_type cache;
    cache.insert({0, 0});
    cache.insert({1, 1});

    // auto helper = [&](unsigned n, cache_type &c)
    std::function<unsigned(unsigned, cache_type&)> helper = [&](unsigned n, cache_type &c)
                      {
                          if (c.count(n) > 0)
                              return c.at(n);
                          c.insert({n, helper(n - 1, c) + helper(n - 2, c)});
                          return c.at(n);
                      };

    return helper(n, cache);
}

unsigned fib3(unsigned n) {
    cache_type cache;
    cache.insert({0, 0});
    cache.insert({1, 1});

    std::function<unsigned(unsigned, cache_type&)> helper = [&](unsigned n, cache_type &c)
                      {
                          if (c.count(n) == 0)
                              c.insert({n, helper(n - 1, c) + helper(n - 2, c)});
                          return c.at(n);
                      };

    return helper(n, cache);
}

unsigned fib5(unsigned n) {
    cache_type cache;
    cache.insert({0, 0});
    cache.insert({1, 1});
    for (unsigned i = 2; i < n + 1; i++)
        cache.insert({i, cache.at(i - 1) + cache.at(i - 2)});
    return cache.at(n);
}

unsigned fib6(unsigned n) {
    if (n < 2)
        return n;
    unsigned a = 0;
    unsigned b = 1;
    unsigned c;
    for (unsigned i = 2; i < n + 1; i++) {
        c = a + b;
        a = b;
        b = c;
    }
    return c;
}

const unsigned F20 = 6765;

TEST(test_fib, fib1) {
    EXPECT_EQ(fib1(20), F20);
}
TEST(test_fib, fib2_nolambda) {
    EXPECT_EQ(fib2_nolambda(0), 0);
    EXPECT_EQ(fib2_nolambda(1), 1);
    EXPECT_EQ(fib2_nolambda(20), F20);

}
TEST(test_fib, fib2) {
    EXPECT_EQ(fib2(0), 0);
    EXPECT_EQ(fib2(1), 1);
    EXPECT_EQ(fib2(20), F20);
}
TEST(test_fib, fib3) {
    EXPECT_EQ(fib3(0), 0);
    EXPECT_EQ(fib3(1), 1);
    EXPECT_EQ(fib3(20), F20);
}
TEST(test_fib, fib5) {
    EXPECT_EQ(fib5(0), 0);
    EXPECT_EQ(fib5(1), 1);
    EXPECT_EQ(fib5(20), F20);
}
TEST(test_fib, fib6) {
    EXPECT_EQ(fib6(0), 0);
    EXPECT_EQ(fib6(1), 1);
    EXPECT_EQ(fib6(20), F20);
}
